rfcs/20190904-tfx-generic-container-based-component.md

```diff
@@ -0,0 +1,863 @@
+is logged and used for caching executions.
+
+[^1]: ML Metadata is also what enables inter-component communication to realize
+    artifact driven component-specific behavior (such as ExampleValidator and
+    ModelValidator)
+
+In order to make this proposal concrete, let’s consider a motivating example of
+creating a simple 2-step pipeline. The first step generates examples, and the
+second trains a model based on the previously produced examples. The following
+shows some example code for these two steps:
+
+*   `my_example_gen.py`
+
+```python
+import argparse
+import pandas as pd
+
+def convert_and_save(df, file):
+  ...
+
+if __name__ == '__main__':
+
+    parser = argparse.ArgumentParser()
+
+    # URI location of input CSV file
+    parser.add_argument('--input_csv_file', ...)
+    # URI location of output CSV file
+    parser.add_argument('--output_csv_file', ...)
+    # Number of CSV columns in the input file
+    parser.add_argument('--num_csv_columns', ...)
+
+    args = parser.parse_args()
+    arguments = args.__dict__
+
+    df = pd.read_csv(arguments['input_csv_file'],
+                     usecols=range(0, int(arguments['num_csv_columns'])))
+
+    # implementation of business logic to ingest dataset
+    convert_and_save(df, arguments['output_csv_file'])
+
+```
+
+*   `my_trainer.py`
+
+```python
+import argparse
+import pandas as pd
+import sklearn.linear_model import Logisticregression
+
+def load_dataset(file):
+  ...
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+
+    # URI location of input file with pre-processed examples.
+    parser.add_argument('--input_examples', ...)
+    # URI location to output the trained model
+    parser.add_argument('--output_model', ...)
+    # Number of training steps to use.
+    parser.add_argument('--train_steps', ...)
+
+    args = parser.parse_args()
+    arguments = args.__dict__
+
+    x_train, y_train, x_eval, y_eval = load_datset(arguments['input_examples'])
+
+    model = LogisticRegression()
+    model.fit(x_train, y_train, args.train_steps)
+
+    #
+    # ... perform grid search, etc, ...
+    #
+
+    # Write the trained model.
+    joblib.dump(model, argument['output_model'])
+```
+
+Given the two (containerized) applications above, our goal is to chain them into
+a 2-step pipeline which will automatically track metadata related to:
+
+*   **Artifacts**: These include the input CSV file, the output CSV file with
+    training and eval examples, and the final Model file.
+*   **Executions**: These include execution steps `MyExampleGen`, with runtime
+    execution property `num_csv_columns`, and `MyTrainer`, with runtime
+    execution property `train_steps`.
+
+Notably, since `my_example_gen.py` produces *Examples*, we make the pipeline
+understand that it is *Examples*, and record it as such in ML Metadata. Doing so
+will enable downstream components, such as Trainer, to understand that it is
```

Comment:
Is schema being tracked out of band here? Why isn't schema tightly coupled to the data itself? For example, can schema be encoded in the file itself. I believe Avro files for example and lots of other file formats allow attach metadata that includes schema.

If your data is stored in a suitable database/feature store, wouldn't that store track schema? 

If schema is attached to the data itself then given a URI downstream files should know what kind of data they are consuming just by reading appropriate metadata.

Do TFRecord files not support encoding metadata and schema in the file.

Reply:
This is an illustration of how one might design (a customized equivalent of) ExampleGen. 

It is totally an option so that the customized (containerized) ExampleGen to also output `Schema` artifact in addition to `Examples` artifact, specialized for particular physical format that has schema information embedded, and the downstream component to consume the produced `Schema` artifact  in the exactly the same way as it would be produced by the stock SchemaGen, so that SchemaGen is no longer necessary.

The proposed DSL extension opens up a door to make such development possible.